Formalize player API

Kiskae opened the pull request:
Draft of API, discussion plz

Kiskae committed:
First draft of player API

Kiskae committed:
Move agent api to separate package, add interface for gamestate control

Kiskae committed:
Large refactor to support gamestate control:
- main() is now in Bootstrap
- Creation of players is handled by GameMonitor
- Created DefaultGameMonitor based on the previous code
- GameState manages the GameSnapshot and performs most of the calls to GameMonitor

MLLeKander commented:
I have multiple comments here.

## Separated API

There seem to be two main use-cases for the *-api.jar pattern: to make it easy to change between implementations and to ease concurrent development. (http://stackoverflow.com/a/9819775)

The ability to change between implementations would be useful if we had multiple implementations of the game logic, or planned to do so in the future.
However, I don't anticipate this being the case for this project, especially since I anticipate development efforts on the server will stop once we formalize the agent API.

The other benefit of the pattern is to make concurrent development easier.
Changes to the API are documented via the *-api.jar rather than the entire jar.
Rather than having an agent target a specific game version ("this agent is compatible with game-1.11 through game-1.23"), you can target an api version ("this agent is compatible with game-api-1.2").
However, again since we don't plan on changing the game logic after this, I don't think this benefit is worth the additional complexity for us.

## Mutabiliy

At very least, agents will need to be able to create a copy of the game state that won't be mutated by the game logic.
The approach taken here is to make everything the agent sees or returns immutable.
I would propose to re-use the some of the type hierarchy we already have (namely, PhysicsCircle) to reduce duplication.

I don't really feel terribly strongly on this point, except I worry it's a solution in search of a problem, and that it complicates our type hierarchy.

## GameMonitor
- This is somewhat awkwardly named, and seems to mix concerns. Why are the details of the construction in the same class as the onMatch\* methods? Why isn't the UI created from here?
- SwingUI sets the state of KeyboardInputProvider? That seems like a horrific separation of concerns.
- GameLoop doesn't support GameMonitor.

## General Java Enterprise-y-ness
- Do we want to use standard Java URL package naming conventions?
- Do we need to have "*Impl" class names ("BallStateImpl", "SideImpl", etc) if it's already namespaced by the impl package?
- Do we need to take care of versioning? If so, how?

## Other Thoughts

I think ultimately, our philosophies differ in the weighting of "how easy it is to implement" vs "how easy it is to fuck up".
I've taken an iterative approach, whereas this is more functional.
Neither is wrong, and the immutable approach is probably better in an enterprise context.
However, I don't know if it's worth it to adopt these practices in the server.

Kiskae commented:
"Seperated Api" - The separation might be me planning too far ahead and conceptualizing a classloading based method where the agents are built against the API and then loaded by the server, separating them makes sense in that context since it disallows any access to the server internals from the perspective of the agents.

"Mutability" - I wouldn't have an issue reusing the PhysicsCircle for position storage honestly.

"GameMonitor" - I agree that the name is awkward, it should probably be named something 'Supervisor' related. The purpose of the monitor is to provide custom configurations for the InputProviders and allow direct control over the flow of the game loop in case it needs to be ended before the match finishes. (Stall detection), UI only displays the game and does not seem to have any effect on the game itself.
The `KeyboardInputProvider` was a quick hack to make it compatible with the previous implementation, could probably be done better.
"GameLoop doesn't support GameMonitor." ??????

"Java Enterprise" - I've definitely not used the standard naming conventions, but it turns out that you cannot refer to classes in the package root from inside a package.
*Impl is again an artifact from making it compatible with the previous implementation.
Versioning was not a consideration when I made the changes.

Kiskae committed:
Make the choice in output perspective explicit through the class the agent extends
AlwaysLeftInputProvider will ensure the player (self()) is always on the left side of the net
The usual InputProvider can be extended to handle the difference between sides yourself.

Kiskae committed:
Let the monitor decide when to stop the game, defaults to original behavior
